frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -46,185 +45,217 @@
-    if (!candidateNodes.containsKey(reportedNode.getHost())) {
-      LOGGER.logDebug("addCandidateNode: %s", reportedNode);
-      candidateNodes.put(reportedNode.getHost(), reportedNode);
-    } else {
-      Node existNode = candidateNodes.get(reportedNode.getHost());
-      existNode.updateFromReportedNode(reportedNode);
-      LOGGER.logDebug("updateCandidateNode: %s ", existNode);
-    }
+  public synchronized void addNode(NodeReport nodeReport) throws Exception {
+    addNode(Node.fromNodeReport(nodeReport));
   }
 
-  public synchronized void removeCandidateNode(NodeReport nodeReport) throws Exception {
-    removeCandidateNode(Node.fromNodeReport(nodeReport));
-  }
+  private void randomizeNodes() {
 
-  @VisibleForTesting
-  public synchronized void removeCandidateNode(Node reportedNode) {
-    if (candidateNodes.containsKey(reportedNode.getHost())) {
-      LOGGER.logDebug("removeCandidateNode: %s", reportedNode);
-      candidateNodes.remove(reportedNode.getHost());
+    filteredNodes.clear();
+    for (String nodeName : allNodes.keySet()) {
+      filteredNodes.add(nodeName);
     }
-  }
-
-  // Add outstanding requested container request
-  public synchronized void addContainerRequest(ContainerRequest request) throws Exception {
-    addContainerRequest(
-        ResourceDescriptor.fromResource(request.getCapability()),
-        request.getNodes());
-  }
-
-  @VisibleForTesting
-  public synchronized void addContainerRequest(ResourceDescriptor resource, List<String> nodeHosts) {
-    for (String nodeHost : nodeHosts) {
-      if (candidateNodes.containsKey(nodeHost)) {
-        candidateNodes.get(nodeHost).addContainerRequest(resource);
-      } else {
-        LOGGER.logWarning("addContainerRequest: Node is no longer a candidate: %s", nodeHost);
-      }
+    int randomTimes = filteredNodes.size();
+    while (randomTimes-- > 0) {
+      int randomIndex = Math.abs(new Random().nextInt(filteredNodes.size()));
+      filteredNodes.add(filteredNodes.remove(randomIndex));
     }
   }
 
-  // Remove outstanding requested container request
-  public synchronized void removeContainerRequest(ContainerRequest request) throws Exception {
-    removeContainerRequest(
-        ResourceDescriptor.fromResource(request.getCapability()),
-        request.getNodes());
-  }
-
-  @VisibleForTesting
-  public synchronized void removeContainerRequest(ResourceDescriptor resource, List<String> nodeHosts) {
-    for (String nodeHost : nodeHosts) {
-      if (candidateNodes.containsKey(nodeHost)) {
-        candidateNodes.get(nodeHost).removeContainerRequest(resource);
-      } else {
-        LOGGER.logWarning("removeContainerRequest: Node is no longer a candidate: %s", nodeHost);
+  private void filterNodesByNodeLabel(String requestNodeLabel) {
+    if (requestNodeLabel != null) {
+      for (int i = filteredNodes.size(); i > 0; i--) {
+        Set<String> availableNodeLabels = allNodes.get(filteredNodes.get(i - 1)).getLabels();
```

Comment:
Why not just use i to indicate current node index?

Reply:
And other places

---
In reply to: [174696342](https://github.com/Microsoft/pai/pull/318#discussion_r174696342) [](ancestors = 174696342)